atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutorImpl.java

```diff
@@ -82,22 +95,73 @@
         int queryId = preparedResult.getItemId();
 
         List<CellWithTimestamp> result = Lists.newArrayList();
-        Iterator<byte[]> rows = rowsAscending.iterator();
-        while (result.size() < limit && rows.hasNext()) {
-            // We can parallelise this, but would need to collect it carefully to make sure we satisfy the guarantees:
-            // 1. Results are ordered
-            // 2. There are at most <limit> results
-            // 3. We return the first <limit> results
-            // We can ensure this by not applying the limit until after collection, and grouping by row
-            // I think this is ensured later on in the code path?
-            CqlResult cqlResult = queryExecutor.executePrepared(queryId,
-                    ImmutableList.of(ByteBuffer.wrap(rows.next())));
-            result.addAll(CqlExecutorImpl.getCells(CqlExecutorImpl::getCellFromRow, cqlResult));
+
+        List<Future<CqlResult>> futures = new ArrayList<>(rowsAscending.size());
+        for (int i = 0; i < rowsAscending.size(); i++) {
```

Comment:
Do we need to initialize everything to null when we already create it to be the right size?

Reply:
Yep. Setting the `rowsAscending.size()` sets the capacity, but not the size. We rely on being able to `set(integer, future)` down, so we need to fill it up.